### ISSUE TITLE
bug: Syncer deleting downstream namespaces unexpectedly on syncer startup

### ISSUE THREAD

**Issue by mikenairn** [kcp-dev/kcp#2041]
### Describe the bug

Downstream namespaces (the ones on the physical cluster) are being deleted when they shouldn't be. This appears to be caused by the syncer thinking that the upstream workspace (the one in KCP) no longer exists when it checks for it's existence on startup . We have also observed the syncer going into a loop of constantly deleting and re-creating namespaces, although I've been unable to re-produce that locally.

syncer logs:

```
I0921 09:55:48.721618       1 namespace_process.go:77] "deleting downstream namespace because the upstream namespace doesn't exist" reconciler="kcp-workload-syncer-namespace" key="kcp-2d1x3k5i2rlu" namespace.workspace="" namespace.namespace="" namespace.name="kcp-2d1x3k5i2rlu" namespace.apiVersion="v1" upstreamWorkspace="root:kuadrant" upstreamNamespace="default"
...
I0921 09:55:54.447445       1 spec_process.go:223] Created downstream namespace kcp-2d1x3k5i2rlu for upstream namespace root:kuadrant|default
```

### Steps To Reproduce

1. Setup kcp with a single sync target and bind to it in a workspace
3. Scale down the syncer `kubectl --kubeconfig ~/.kube/config --context kind-kcp-cluster-1 scale deployments/kcp-syncer-kcp-cluster-1-4yiou870 --replicas=0 -n kcp-syncer-kcp-cluster-1-4yiou870`
4. Wait for sync targets "Ready" status to be false `kubectl get synctargets -o wide -w`
5. Scale up the syncer `kubectl --kubeconfig ~/.kube/config --context kind-kcp-cluster-1 scale deployments/kcp-syncer-kcp-cluster-1-4yiou870 --replicas=1 -n kcp-syncer-kcp-cluster-1-4yiou870`
6. Observe the kcp namespaces being deleted and re-created `kubectl --kubeconfig ~/.kube/config --context kind-kcp-cluster-1 get ns -w | grep kcp-`

### Expected Behaviour

Syncer should not delete namespaces from the physical cluster when they exist in KCP.

### Additional Context

Observed with KCP 0.8.x, syncer image ghcr.io/kcp-dev/kcp/syncer:v0.8.2

**Comment by jmprusi** [comment]
I think that's the expected behavior if it goes like this:

1. SyncTarget Ready == true (as syncer is running)
2. Resources are scheduled (label) to that SyncTarget, referenced by its SynctargetKey.
3. Syncer is scaled down.
**4. Scheduler detects that the SyncTarget is not Ready, then removes the Synctarget Label on those resources.**
5. Syncer is scaled up.
6. Reconciles current downstream resources, and checks for the upstream resources to exist and be labeled with the syncer synctargetkey == **no upstream resources are labeled, so those not really exist for the syncer** == remove downstream resources.
7. Syncer heartbeats and SyncTarget is marked as Ready
8. Scheduler sees a change in SyncTarget readiness and schedules the resources**.
9. Syncer picks the new upstream resources and syncs those back.

Can you make sure that the issues is not reproducible if you scale down & scale up the syncer quickly, in less than a minute or so? SyncTarget Readiness remains True.

**Comment by jmprusi** [comment]
Another possibly related instance of this issue: 

https://kubernetes.slack.com/archives/C021U8WSAFK/p1666709160236279

**Comment by Roming22** [comment]
We're strongly impacted by this bug, which results in our service to be highly unstable as it triggers another bug related to finalizers. Unfortunately we can't expect that second bug to be resolved in a timely manner, so we're hoping for a fix on kcp's end.

**Comment by davidfestal** [comment]
This should be fixed on main. Now downstream namespaces are only deleted when no more synced resources are inside them, and even then, only after a 30 s delay (configurable).

**Comment by davidfestal** [comment]
@Roming22 @mikenairn Can you confirm that it is fixed ?

**Comment by ncdc** [comment]
@jmprusi you indicated in #2264 that it was only a partial fix for this issue. @davidfestal you wrote above this is fixed on main. Was there another PR that fully fixed this?
